contracts/utils/cryptography/RSA.sol

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {Math} from "../math/Math.sol";
+
+/**
+ * @dev RSA PKCS#1 v1.5 signature verification implementation according to https://datatracker.ietf.org/doc/html/rfc8017[RFC8017].
+ *
+ * This library supports PKCS#1 v1.5 padding to avoid malleability via chosen plaintext attacks in practical implementations.
+ * The padding follows the EMSA-PKCS1-v1_5-ENCODE encoding definition as per section 9.2 of the RFC. This padding makes
+ * RSA semanticaly secure for signing messages.
+ *
+ * Inspired by https://github.com/adria0/SolRsaVerify[Adrià Massanet's work]
+ */
+library RSA {
+    /**
+     * @dev Same as {pkcs1} but using SHA256 to calculate the digest of `data`.
+     */
+    function pkcs1Sha256(
+        bytes memory data,
+        bytes memory s,
+        bytes memory e,
+        bytes memory n
+    ) internal view returns (bool) {
+        return pkcs1(sha256(data), s, e, n);
+    }
+
+    /**
+     * @dev Verifies a PKCSv1.5 signature given a digest according the verification
+     * method described in https://datatracker.ietf.org/doc/html/rfc8017#section-8.2.2[section 8.2.2 of RFC8017].
+     *
+     * IMPORTANT: Although this function allows for it, using n of length 1024 bits is considered unsafe.
+     * Consider using at least 2048 bits.
+     *
+     * @param digest the digest to verify
+     * @param s is a buffer containing the signature
+     * @param e is the exponent of the public key
+     * @param n is the modulus of the public key
+     */
+    function pkcs1(bytes32 digest, bytes memory s, bytes memory e, bytes memory n) internal view returns (bool) {
+        unchecked {
+            // cache and check length
+            uint256 length = n.length;
+            if (
+                length < 0x40 || // PKCS#1 padding is slightly less than 0x40 bytes at the bare minimum
+                length != s.length // signature must have the same length as the finite field
+            ) {
+                return false;
+            }
+
+            // RSAVP1 https://datatracker.ietf.org/doc/html/rfc8017#section-5.2.2
+            (bool success, bytes memory buffer) = Math.tryModExp(s, e, n);
+            if (!success) {
+                return false;
+            }
```

Comment:
tryModExp doesn't do that check, and we have nothing that guarantee that s < n.

We should probably implement a bytes comparaison function, and use it here.

Reply:
we could use something like 

```
enum CmpResult { Eq, Lt, Gt }

function compare(bytes memory a, bytes memory b) pure returns (CmpResult) {
    unchecked {
        uint256 i = 0; // pointer on the a buffer
        uint256 j = 0; // pointer on the b buffer
        if (a.length > b.length) {
            while (i < a.length - b.length) {
                if (a[i++] != 0) return CmpResult.Gt;
            }
        } else if (a.length < b.length) {
            while (j < b.length - a.length) {
                if (b[j++] != 0) return CmpResult.Lt;
            }
        }
        while (i < a.length) { // eq j < b.length
            bytes1 ai = a[i++];
            bytes1 bj = b[j++];
            if (ai < bj) {
                return CmpResult.Lt;
            } else if (ai > bj) {
                return CmpResult.Gt;
            }
        }
        return CmpResult.Eq;
    }
}
```

Reply:
@frangio 
I think we need this to avoid signature maleability ... but it also feels like an expensive check. WDYT?

We could write it in assembly ...
I'm going to check for other existing implementation

Reply:
or we could check that 
```solidity
keccak256(s) == keccak256(modExp(s, 1, n))
```

Reply:
I'd be fine with `keccak256(s) == keccak256(modExp(s, 1, n))`.

`modExp` is only 200 gas if `e == 1`, and the cost of both keccak256 is 30.
The bottom line is that we're adding ~300 gas units, which is about 12 cents at 100gwei and ETH at 4000. 
I think this is good enough, and the loop might be even more costly if we end up accepting 4096-bit keys.

Reply:
There is also the cost of the two keccak. I suspect the loop might be cheaper (particularly is we do some fancy assembly that checks in blocks of 32 bytes)

Reply:
Right, fair points. I also didn't consider under normal circumstances, the `compare` function finishes as early as possible.

If we're adding the `compare` function, I'd put it in `Math.sol`, wdyt?

Reply:
Note that we have already validated that `s.length == n.length` above, do you think the following would work?

```
enum CmpResult { Eq, Lt, Gt }
function compare(bytes memory a, bytes memory b) pure returns (CmpResult) {
    uint256 higherA = 0;
    uint256 higherB = 0;
    assembly ("memory-safe") {
        higherA := mload(add(a, 0x20))
        higherB := mload(add(b, 0x20))
    }
    return
        higherA == higherB
            ? CmpResult.Eq
            : higherA < higherB
                ? CmpResult.Lt
                : CmpResult.Gt;
}
```

or at least:

```
uint256 i = 0; // pointer on the a buffer
uint256 j = 0; // pointer on the b buffer
if (a.length > b.length) {
    while (i < a.length - b.length) {
        if (a[i++] != 0) return CmpResult.Gt;
    }
} else if (a.length < b.length) {
    while (j < b.length - a.length) {
        if (b[j++] != 0) return CmpResult.Lt;
    }
}
while (i < a.length) { // eq j < b.length
    bytes1 ai = a[i++];
    bytes1 bj = b[j++];
    if (ai < bj) {
        return CmpResult.Lt;
    } else if (ai > bj) {
        return CmpResult.Gt;
    }
}
return CmpResult.Eq;
```

Reply:
first option only checks the first 32 bytes ... if the difference is further its not catching it. Arbitrary length means we need some loop somewhere.

Reply:
worst case in the first option is that they're 32-bytes equal, would it be possible to just require an explicit `Gt` and revert otherwise?

Reply:
https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4952/commits/abe598d94f12e51648b9ea0e741083043e834e92

Reply:
Do we know what would be the average cost of this check?

Reply:
The for could be long (64 iterations for a 4096 key)  ... but there is a break !

All real cases should have a difference somewhere in the first 32bytes (having such a large equality is not realisitc IMO).

One iteration should be way under 100 gas.

Reply:
>64 iterations for a 4096 key

Are you sure about this? Isn't it 4096 bits, so 512 bytes, which is 16 32-byte words.